Drop all capabilities from control-plane containers

michaelbannister opened the pull request:
As suggested by @rlenglet during discussion of #20311. These containers don't run as root
and don't need any capabilities enabled.

michaelbannister committed:
Drop all capabilities from control-plane containers

As suggested during discussion of #20311. These containers don't run as root
and don't need any capabilities enabled.

istio-testing commented:
Hi @michaelbannister. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

howardjohn commented:
/ok-to-test

istio-testing commented:
@michaelbannister: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | 88e6774a859c7f1a2db3f1914520b22a06347514 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20854/integ-distroless-k8s-tests_istio/935) | `/test integ-distroless-k8s-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@michaelbannister: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | 88e6774a859c7f1a2db3f1914520b22a06347514 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20854/integ-distroless-k8s-tests_istio/936) | `/test integ-distroless-k8s-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

michaelbannister commented:
failed to create containerd task: OCI runtime create failed: container_linux.go:346: starting container process caused \"chdir to cwd (\\\"/home/nonroot\\\") set in config.json failed: permission denied\": unknown

michaelbannister commented:
`istio-testing/citadel:istio-testing-distroless` failed starting:
```
failed to create containerd task: OCI runtime create failed: container_linux.go:346: starting container process caused \"chdir to cwd (\\\"/home/nonroot\\\") set in config.json failed: permission denied\": unknown
```

rlenglet reviewed (approved):
Nice, thanks!

rlenglet commented:
```
2020-02-04T21:45:32.188040Z	info	ci	Checking pods ready...
2020-02-04T21:45:32.192804Z	info	ci	  [ 0]                istio-citadel-7d97bbcf88-krkb7         Running (container not ready: 'citadel')
```

michaelbannister commented:
The distroless base image has this:
```
Permission     UID:GID       Size  Filetree
drwxr-xr-x         0:0     1.8 MB  └── .
drwxr-xr-x         0:0        0 B      ├── bin
drwxr-xr-x         0:0        0 B      ├── boot
drwxr-xr-x         0:0        0 B      ├── dev
drwxr-xr-x         0:0     259 kB      ├─⊕ etc
drwxr-xr-x 65532:65532        0 B      ├── home
drwx------ 65532:65532        0 B      │   └── nonroot
```

…and trying to run just this base image as user:group 1337:1337 gives the same error:
```
$ docker run --rm -it -u 1337:1337 -w /home/nonroot gcr.io/distroless/static:nonroot /bin/sh
docker: Error response from daemon: OCI runtime create failed: container_linux.go:346: starting container process caused "chdir to cwd (\"/home/nonroot\") set in config.json failed: permission denied": unknown.
```

istio-testing commented:
@michaelbannister: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | 88e6774a859c7f1a2db3f1914520b22a06347514 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20854/integ-distroless-k8s-tests_istio/936) | `/test integ-distroless-k8s-tests_istio`
integ-security-k8s-tests_istio | 88e6774a859c7f1a2db3f1914520b22a06347514 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20854/integ-security-k8s-tests_istio/7904) | `/test integ-security-k8s-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

michaelbannister commented:
But I don't understand why this is failing now when it didn't for my previous PR…

rlenglet commented:
Interesting. Why are we relying on the home directory of that user...

howardjohn commented:
We need to run in `/`, we rely on it to support VMs/development otherwise it tries to write to things we don't have access to

michaelbannister commented:
> Interesting. Why are we relying on the home directory of that user...

If you look at the [distroless image base](https://github.com/GoogleContainerTools/distroless/blob/master/base/base.bzl#L77) it's the equivalent of setting `WORKDIR /home/nonroot` and that directory has those permissions I showed above. When we specify a UID/GID to the container runtime it seems to try and change directory to /home/nonroot and fails because UID 1337 doesn't have permission?

But AFAICT this hasn't changed recently (including in the base image?)

howardjohn commented:
Do we just need something like https://github.com/istio/istio/blob/0996b80fd41dc9ac92e6d2129a78a83c68aff76a/pilot/docker/Dockerfile.proxyv2#L46?

rlenglet commented:
> Do we just need something like https://github.com/istio/istio/blob/0996b80fd41dc9ac92e6d2129a78a83c68aff76a/pilot/docker/Dockerfile.proxyv2#L46?

+1 The `WORKDIR` may be inherited from the base image.

michaelbannister commented:
I could certainly change the WORKDIR – but what I don't understand is why this failure should show up on this change, where we're just dropping capabilities, rather than my previous change which made the containers run as 1337:1337 ?

rlenglet commented:
Here are Docker's default capabilities that we're dropping here:
https://github.com/moby/moby/blob/46ec8731fbce9b13e41bd85d6cea691406bba852/oci/caps/defaults.go#L6-L19
```
		"CAP_CHOWN",
		"CAP_DAC_OVERRIDE",
		"CAP_FSETID",
		"CAP_FOWNER",
		"CAP_MKNOD",
		"CAP_NET_RAW",
		"CAP_SETGID",
		"CAP_SETUID",
		"CAP_SETFCAP",
		"CAP_SETPCAP",
		"CAP_NET_BIND_SERVICE",
		"CAP_SYS_CHROOT",
		"CAP_KILL",
		"CAP_AUDIT_WRITE",
```
Maybe we're changing file ownerships somewhere in the code?

michaelbannister commented:
I think it's not getting as far as running the entrypoint because the UID/GID doesn't allow to change to the workdir.

I built this very minimal image:
```
$ docker build -q -t test-workdir-permissions - <<EOF
FROM tianon/true as source
FROM gcr.io/distroless/static:nonroot
COPY --from=source /true /true
ENTRYPOINT ["/true"]
EOF
```
and then ran it:
```
$ docker run --rm -it -u 1337:1337 test-workdir-permissions
docker: Error response from daemon: OCI runtime create failed: container_linux.go:346: starting container process caused "chdir to cwd (\"/home/nonroot\") set in config.json failed: permission denied": unknown.
$ docker run --rm -it -u 1337:1337 -w / test-workdir-permissions
[works just fine]
```

So – yes, we could change the workdir in the dockerfile. But I still don't get why this was working _before_ my change, and not working once we explicitly drop capabilities?

michaelbannister commented:
I won't have a chance to do any more on this until Sunday at earliest. If anyone else is keen to pick it up and submit a replacement PR, I shan't be offended!

michaelbannister commented:
OK, so this appears to be a difference in behaviour between Kind and Kubernetes proper.
Details here for a simple reproducible demonstration:
https://github.com/michaelbannister/distroless-permissions-test

This could mean that the changes in #20311 mean that deployment of the _distroless_ citadel (and maybe other images?) won't actually work in a full Kubernetes cluster? :(

michaelbannister committed:
Set workdir to / for distroless-based images

The distroless base has workdir as /home/nonroot but we run as 1337:1337
which has no permissions on that directory so the runtime fails to start
the container. The standard base has workdir / so change to match that.

michaelbannister commented:
This now fixes the problem with the `-distroless` images introduced in #20311 but only discovered here.
This should either make this a candidate for cherrypick into release-1.5, or else #20820 should probably be reverted from release-1.5.

sdake commented on pilot/docker/Dockerfile.pilot:
```diff
@@ -12,6 +12,7 @@ USER 1337:1337
 # The following section is used as base image if BASE_DISTRIBUTION=distroless
 # hadolint ignore=DL3007
 FROM gcr.io/distroless/static:nonroot as distroless
+WORKDIR /
```
redundant `WORKDIR /` results in a small ineffeciency.

michaelbannister replied:
It's not redundant at all: it means this image will actually run.

My previous PR (#20311) changed the Deployments so that the containers run as user/group 1337:1337.
Kubernetes (or Docker on its own) can't successfully run a container from this image when setting user/group to 1337:1337 – see [this comment](#issuecomment-582148470) – because the distroless base image has set `WORKDIR /home/nonroot` and user 1337 has no permissions to reach that directory. By changing the workdir to `/` we fix this problem.

Really, this fix should have gone in #20311 but I didn't discover this problem then due to a difference in behaviour with kind compared to Kubernetes. Kind only showed up the same error once I'd configured the container's security context to drop all capabilities (i.e. this PR). I'm planning to raise an issue on the kind project shortly.

rlenglet replied:
I think @sdake's point is that we're already doing this 2 lines below in this particular `Dockerfile`. It's still required in the other `Dockerfile`s, I don't think there's any question about that.

michaelbannister replied:
Ah, my apologies @sdake – the minimal context shown in the PR view meant I didn't see this! Will remove this line & push.

sdake replied:
@michaelbannister - thanks. I didn't realize the context wasn't shown or I would have offered more information in the review. Using GitHub is a bit robotic for me 😁

michaelbannister committed:
Remove duplicate WORKDIR directive

istio-testing commented:
@michaelbannister: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-istioio-k8s-tests_istio | a42f2b47f82461906e6c2ced9aa081a3194414bb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20854/integ-istioio-k8s-tests_istio/4792) | `/test integ-istioio-k8s-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

michaelbannister commented:
/test integ-istioio-k8s-tests_istio

michaelbannister commented:
/cherry-pick release-1.5

istio-testing commented:
@michaelbannister: only [istio](https://github.com/orgs/istio/people) org members may request cherry-picks. You can still do the cherry-pick manually.

<details>

In response to [this](https://github.com/istio/istio/pull/20854#issuecomment-585350806):

>/cherry-pick release-1.5


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

rlenglet commented:
/cherry-pick release-1.5

istio-testing commented:
@rlenglet: once the present PR merges, I will cherry-pick it on top of release-1.5 in a new PR and assign it to you.

<details>

In response to [this](https://github.com/istio/istio/pull/20854#issuecomment-585359418):

>/cherry-pick release-1.5


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

BenTheElder commented:
> OK, so this appears to be a difference in behaviour between Kind and Kubernetes proper.

Nitpicking but, KIND _is_ Kubernetes proper, we ship fully upstream unmodified kubernetes :+)

This appears to be a difference in containerd and dockerd with `WORKDIR` (?), I've looped in a maintainer there to look further after a bit of debug myself,.

We can reproduce the same behavior on GKE with containerd versus dockerd nodes with the internal `chdir` failing on docker or containerd+the cap drop all but not containerd without the cap drop.

Setting the WORKDIR explicitly seems like a sane approach in the meantime imho.

BenTheElder commented:
Upstream fix pending in https://github.com/containerd/cri/pull/1397, we'll pull it in once it's merged.

michaelbannister commented:
Still awaiting review approval from @costinm @linsun @rshriram (I'm unclear whether this is OR or AND). This also needs cherry-picking to fix a bug in release 1.5, which can't happen until this is merged to master.

howardjohn commented:
you need to look at the bottom of the pr not the reviews list - "Waiting on code owner review from istio/wg-policies-and-telemetry-maintainers and/or istio/wg-security-maintainers". You still have the ambiguous and/or though... in this case I happen to knows it's and (based on CODEOWNERS file).

mandarjog reviewed (approved):
/lgtm

istio-testing commented:
@rlenglet: #20854 failed to apply on top of branch "release-1.5":
```Applying: Drop all capabilities from control-plane containers
Using index info to reconstruct a base tree...
M	manifests/istio-control/istio-config/templates/deployment.yaml
M	manifests/istio-control/istio-discovery/templates/deployment.yaml
M	manifests/istio-policy/templates/deployment.yaml
M	manifests/istio-telemetry/mixer-telemetry/templates/deployment.yaml
M	manifests/security/citadel/templates/deployment.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/all_on.golden-show-in-gh-pull-request.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/component_hub_tag.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/flag_force.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/flag_output.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/flag_output_set_profile.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/flag_output_set_values.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/flag_override_values.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/flag_set_values.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/pilot_default.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/pilot_k8s_settings.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/pilot_override_kubernetes.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/pilot_override_values.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/telemetry_default.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/telemetry_k8s_settings.golden.yaml
A	operator/cmd/mesh/testdata/manifest-generate/output/telemetry_override_kubernetes.golden.yaml
M	operator/pkg/vfs/assets.gen.go
Falling back to patching base and 3-way merge...
Auto-merging operator/pkg/vfs/assets.gen.go
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/telemetry_override_kubernetes.yaml
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/telemetry_k8s_settings.yaml
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/telemetry_default.yaml
CONFLICT (modify/delete): operator/cmd/mesh/testdata/manifest-generate/output/pilot_override_values.golden.yaml deleted in HEAD and modified in Drop all capabilities from control-plane containers. Version Drop all capabilities from control-plane containers of operator/cmd/mesh/testdata/manifest-generate/output/pilot_override_values.golden.yaml left in tree.
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/pilot_override_kubernetes.yaml
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/pilot_k8s_settings.yaml
CONFLICT (content): Merge conflict in operator/cmd/mesh/testdata/manifest-generate/output/pilot_k8s_settings.yaml
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/pilot_default.yaml
CONFLICT (content): Merge conflict in operator/cmd/mesh/testdata/manifest-generate/output/pilot_default.yaml
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/flag_set_values.yaml
CONFLICT (content): Merge conflict in operator/cmd/mesh/testdata/manifest-generate/output/flag_set_values.yaml
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/flag_override_values.yaml
CONFLICT (content): Merge conflict in operator/cmd/mesh/testdata/manifest-generate/output/flag_override_values.yaml
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/flag_output_set_values.yaml
CONFLICT (content): Merge conflict in operator/cmd/mesh/testdata/manifest-generate/output/flag_output_set_values.yaml
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/flag_output_set_profile.yaml
CONFLICT (content): Merge conflict in operator/cmd/mesh/testdata/manifest-generate/output/flag_output_set_profile.yaml
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/flag_output.yaml
CONFLICT (content): Merge conflict in operator/cmd/mesh/testdata/manifest-generate/output/flag_output.yaml
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/flag_force.yaml
CONFLICT (content): Merge conflict in operator/cmd/mesh/testdata/manifest-generate/output/flag_force.yaml
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/component_hub_tag.yaml
CONFLICT (content): Merge conflict in operator/cmd/mesh/testdata/manifest-generate/output/component_hub_tag.yaml
Auto-merging operator/cmd/mesh/testdata/manifest-generate/output/all_on.yaml
CONFLICT (content): Merge conflict in operator/cmd/mesh/testdata/manifest-generate/output/all_on.yaml
Auto-merging manifests/security/citadel/templates/deployment.yaml
Auto-merging manifests/istio-telemetry/mixer-telemetry/templates/deployment.yaml
Auto-merging manifests/istio-policy/templates/deployment.yaml
Auto-merging manifests/istio-control/istio-discovery/templates/deployment.yaml
Auto-merging manifests/istio-control/istio-config/templates/deployment.yaml
error: Failed to merge in the changes.
Patch failed at 0001 Drop all capabilities from control-plane containers

```

<details>

In response to [this](https://github.com/istio/istio/pull/20854#issuecomment-585359418):

>/cherry-pick release-1.5


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

michaelbannister commented:
I'm preparing a separate PR to fix the cherry-pick into release-1.5. Will link here when it's ready.

michaelbannister commented:
Note: the merge conflicts were in the generated-but-committed files under `operator/cmd/mesh/testdata/manifest-generate/output`.
